### ISSUE TITLE
Ping - Notification Delivery Service

### ISSUE THREAD

**Issue by Saumay-Agrawal** [onflow/developer-grants#167]
# Ping

# Grant category

- Developer tools / services

# Description

## Problem statement

At present, there’s no direct communication channel between dapp developers and their respective users. Users have to rely on public platforms like Discord, Twitter, Medium etc.

- These platforms have a lot of noise and very low signal.
- And users also need to keep a check on different social accounts/servers for different dapps.

This makes it difficult for users to stay updated with relevant information, and there's a chance of missing out on critical updates & announcements. 

- Target audience
   - Dapp developers
   - Dapp Users

- Evidence for the need
   - Layer-1 ecosystems benefit heavily from such platforms. It creates a direct connection between Dapp developers and their users. This leads to higher engagement with the applications. On other L1s, similar platforms have received a very positive response and have been integrated with almost all major dapps of that ecosystem. For eg
      - [Push](https://push.org/) on Ethereum
      - [Dialect](https://www.dialect.to/) on Solana
   - As Flow’s target audience is general public, it would benefit hugely from this. For example, NFT All Day users can get direct updates in their inbox. They would not need to scrape through Twitter accounts, Discord servers etc for updates.

## Proposed solution

There should be a notification service that creates a direct communication channel between dapp developers and their respective users.

On one hand, the platform will allow Dapp developers to

- login/authenticate on the platform via wallet
- create a publisher profile
- and send updates to their subscribers

On the other hand, it will enable dapp users to 

- login/authenticate on the platform via wallet
- view list of publishers (dapps)
- subscribe to a dapp for updates
- view updates from all dapps they have subscribed to at one place

I developed a very basic but functional version of its core architecture layer in the recent Flow Hackathon. You can check it out here: [https://devfolio.co/projects/ping-107f](https://devfolio.co/projects/ping-107f)

## Impact

TLDR; Overall better application engagement across Flow ecosystem. 

- Ping will reduce the information asymmetry between dapp developers and dapp users.
- It creates a feedback loop so that dapp users stay updated about the key activities happening or about to happen on the dapp.
- This will reduce the information overload on users by a huge magnitude and will lead higher engagement with applications.
- For example, it can improve governance participation by keeping users in loop of various governance related activites. It can also lead to higher engagement in gaming applications. It can keep an NFT community engaged by updating community members about ongoing/upcoming events. 

Note - This a proof-of-concept. If we get a good response from Flow community, there’s we can also add some major upgrades, like a mobile application with push notifications, integration with various services like .find & flowNS to make user experience even better etc.

# Milestones and funding

> Note: Please consider adoption and/or maintenance milestones at the end of your project.

<table class="tg">
<thead>
  <tr>
    <th class="tg-0lax">Milestone</th>
    <th class="tg-0lax">Deliverables</th>
    <th class="tg-0lax">Timelines</th>
    <th class="tg-0lax">Risk</th>
    <th class="tg-0lax">USD Proposal</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-0lax">1 - Dapp developer flow</td>
    <td class="tg-0lax">UI allowing dapp developers to<br>- login (wallet integration flow)<br>- register as publishers (profile setup)<br>- send updates to their subscribers<br><br>Relevant functionalities deployed as smart contracts on testnet </td>
    <td class="tg-0lax">4 weeks<br><br>(~150 hours FTE)</td>
    <td class="tg-0lax">-</td>
    <td class="tg-0lax">$15,000</td>
  </tr>
  <tr>
    <td class="tg-0lax">2 - Dapp User flow</td>
    <td class="tg-0lax">UI allowing dapp users to<br>- login (wallet integration flow)<br>- view publishers<br>- subscribe to a publisher<br>- view messages from subscribed publishers<br><br>Relevant functionalities deployed as smart contracts on testnet</td>
    <td class="tg-0lax">4 weeks<br><br>(~150 hours FTE)</td>
    <td class="tg-0lax">-</td>
    <td class="tg-0lax">$15,000</td>
  </tr>
  <tr>
    <td class="tg-0lax">3 - Alpha launch</td>
    <td class="tg-0lax">Deploy application on mainnet<br><br>Onboard Dapp developers and users<br><br>Take feedback, resolve bugs if found any</td>
    <td class="tg-0lax">4 weeks<br><br>(~70 hours FTE)</td>
    <td class="tg-0lax">-</td>
    <td class="tg-0lax">$7,000 </td>
  </tr>
</tbody>
</table>

Total funding proposed: $37,000

Note - For this project, this includes infra costs & post-completion support in the form of minor updates and bug fixes for the duration of 6 months post grant completion.
  
# Team

<table class="tg">
<thead>
  <tr>
    <th class="tg-0pky">Name</th>
    <th class="tg-0pky">Role</th>
    <th class="tg-0pky">Bio</th>
    <th class="tg-0pky">Contact</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-0pky">Saumay Agrawal</td>
    <td class="tg-0pky">Full-stack Software Engineer</td>
    <td class="tg-0pky">- 6+ years of experience in software development across different product teams, startups &amp; research labs<br>- 3+ years of experience in building web3 applications for multiple clients and open source orgs like Filecoin, Livepeer, Celo etc. Some examples:<br>- <a href="https://github.com/buidl-labs/Demux" target="_blank" rel="noopener noreferrer">Demux</a>: a gateway for decentralised streaming ecosystem using Livepeer and Filecoin<br>- <a href="https://churrito.fi/" target="_blank" rel="noopener noreferrer">ChurroFi</a>: Voting (Staking) platform for Celo<br>- LinkedIn: <a href="https://www.linkedin.com/in/saumayagrawal/" target="_blank" rel="noopener noreferrer">https://www.linkedin.com/in/saumayagrawal/</a></td>
    <td class="tg-0pky"><a href="mailto:hi@saumay.com" target="_blank" rel="noopener noreferrer">hi@saumay.com</a></td>
  </tr>
</tbody>
</table>

**Comment by joshuahannan** [comment]
I'm not in charge of the developer grants so I can't provide the final decision, but I'll offer a little bit of feedback. This sounds like a really interesting idea that could be really useful, but I don't know if it fits into the developer tools category. This seems more like a product that you could offer to teams in the Flow ecosystem. I believe that grants are more focused towards fundamental improvements to the developer experience, so I don't know if this will count towards that, but @chrisackermann will be able to provide more feedback. 👍

**Comment by franklywatson** [comment]
Hi @Saumay-Agrawal - thanks for submitting this grant proposal. We are presently making changes to the grant program and application process and as a result have not been able to review this. Once the grant program submissions are opened up again we will follow up and let you know next steps. Thanks for your patience.

**Comment by Aliserag** [comment]
We appreciate your proposal. After careful review we've decided not to proceed at this time. Please stay connected to our Twitter and working groups to stay on top of the latest developments in the ecosystem and new grant opportunities.
